src/sys/unix/selector/poll.rs

```diff
@@ -0,0 +1,725 @@
+        let mut closed_raw_fds = Vec::new();
+
+        loop {
+            // Complete all current operations.
+            loop {
+                if self.waiting_operations.load(Ordering::SeqCst) == 0 {
+                    break;
+                }
+
+                fds = self.operations_complete.wait(fds).unwrap();
+            }
+
+            // Perform the poll.
+            trace!("Polling on {:?}", &fds);
+            let num_events = poll(&mut fds.poll_fds, deadline)?;
+            if num_events == 0 && deadline.map(|v| v <= Instant::now()).unwrap_or(false) {
+                // timeout
+                return Ok(());
+            }
+
+            trace!("Poll finished: {:?}", &fds);
+            let notified_events = fds.poll_fds[0].0.revents;
+            let notified = notified_events != 0;
+            let mut num_fd_events = if notified { num_events - 1 } else { num_events };
+
+            let mut pending_wake_tokens_guard = self.pending_wake_tokens.lock().unwrap();
+            let pending_wake_tokens =
+                std::mem::replace(pending_wake_tokens_guard.as_mut(), Vec::new());
+            drop(pending_wake_tokens_guard);
+
+            if notified {
+                self.notify_waker.ack_and_reset();
+                num_fd_events += pending_wake_tokens.len();
+            }
+
+            // We now check whether this poll was performed with descriptors which were pending
+            // for removal and filter out any matching.
+            let mut pending_removal_guard = self.pending_removal.lock().unwrap();
+            let mut pending_removal = std::mem::replace(pending_removal_guard.as_mut(), Vec::new());
+            drop(pending_removal_guard);
+
+            // Store the events if there were any.
+            if num_fd_events > 0 {
+                let fds = &mut *fds;
+
+                events.reserve(num_fd_events);
+
+                for pending_wake_token in pending_wake_tokens {
+                    events.push(Event {
+                        token: pending_wake_token,
+                        events: notified_events,
+                    });
+                }
+
+                for fd_data in fds.fd_data.values_mut() {
+                    let PollFd(poll_fd) = &mut fds.poll_fds[fd_data.poll_fds_index];
+
+                    if pending_removal.contains(&poll_fd.fd) {
+                        // Fd was removed while poll was running
+                        continue;
+                    }
+
+                    if poll_fd.revents != 0 {
+                        // Store event
+                        events.push(Event {
+                            token: fd_data.token,
+                            events: poll_fd.revents,
+                        });
+
+                        if poll_fd.revents & (libc::POLLHUP | libc::POLLERR) != 0 {
+                            pending_removal.push(poll_fd.fd);
+                            closed_raw_fds.push(poll_fd.fd);
+                        }
```

Comment:
So there's a few ways to do this but at a high level the problem to solve is that with edge triggered behavior once an fd is closed or shutdown you won't get spammed with err/hup events in a tight loop; with level triggered you will.  So once you receive a terminal event from poll you have to internally track that you shouldn't call poll anymore with that fd (removing events flags isn't enough) but let external callers still safely call deregister since ofc from their POV it is still registered.

In effect the solution I went with (which is also how the Windows Selector seems to do it) is that but it turns deregister into a two layered concept: the lowest layer removes the fd entirely and just forgets it exists, while the higher layer uses an AtomicBoolean to stay aligned with the lower layer and no op calls to deregister.

There are alternatives though, such as simply removing the fds from the list of fds sent to poll but still keeping FdData around with a new flag indicating it's closed.  Then when you deregister it just ignores that it isn't in the poll_fds set.  I like this solution generally but chose not to go there as I think it's slightly more fragile as there's already quite a few edge cases in the poll loop to keep track of.  Additionally there was an optimization I avoided (you can modify the internal data structures while holding the lock inside the main poll loop) simply to avoid making that code even harder to follow / more likely to break.

I'm interested in feedback here though and happy to pivot, just wanted to help you understand my thinking so far.

Reply:
That makes sense... I don't really have a good intuition into what option would be best, so instead let me throw another idea in the ring.

Instead of removing the fd, we can also set the fd to a negative number (i.e. `poll_fd.fd -= poll_fd.fd`), then it will be ignored by the kernel. A little less overhead for us, but a little more for the kernel...

But as I said, I'm not sure what the best option is.

I'm fine with any of them really, they all make sense to me, but a comment explaining the situation would be nice.

Reply:
Ack'd, I'll add some language around why we're removing before we actually deregister from the outside.